300 feature finish question page requests

rylan-geykhman opened the pull request:
<!--------------------------------
Before you submit your PR:
1. Do you have any console.logs for testing purposes? Remove them please!
2. Does your code actually work? Test all of your changes one more time!
3. Fill out the section below. It helps us review your PR because I ain't reading allat lil bro
4. Delete these instruction comments (optional)
--------------------------------->

<!--------------------------------------------------------------------------------------------------------------
Summarize any changes your PR includes. If they are related to an issue, tag it at the end.

Examples:
- Added yoda to home page
- Created yoda clicker component (#69420)
- Fixed bug where yoda does not spawn after 5 clicks (#80085)
-------------------------------------------------------------------------------------------------------------->

### Changes

- Questions page now made.
- Supports both dynamic and static questions.
- You can answer questions / submit assignments.

rylan-geykhman committed:
commit

rylan-geykhman committed:
fetch stuff

- testing why getNextQuestion doesn't work at the moment
- trying to display question details (text and answer choices)
- after implement checking "isStatic" for the assignment

rylan-geykhman committed:
commit

- commented out something
- dynamic question = random questions, cant move, deltamath
- static question = no random questions, like ap classroom

rylan-geykhman committed:
Merge branch 'dev' into 300-feature-finish-question-page-requests

rylan-geykhman committed:
commit?

- now differentiates between static / dynamic
- need to disable and enable buttons dependent on static / dynamic
- tags on ends of questions, function for images to work
- on pace?

rylan-geykhman committed:
static / dynamic disabling

- disabled back and next buttons for dynamic questions
- need to also disable sidebar
- images need to load
- submitting questions for dynamic questions...

rylan-geykhman committed:
some changes

- static questions have an issue with the link (q=0 takes to q1, then clicking next makes q=1 in the link and is still the same question 1)
- answers that are chosen need to be stored
- dynamic done, but when leaving assignment the question index doesnt save.

rylan-geykhman committed:
commit

rylan-geykhman committed:
Merge branch 'dev' into 300-feature-finish-question-page-requests

rylan-geykhman committed:
error message

rylan-geykhman committed:
commit

rylan-geykhman committed:
cleaning up + comments

rylan-geykhman committed:
commit

rylan-geykhman committed:
commit

rylan-geykhman committed:
removing comments

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -29,6 +30,15 @@
                 </button>
               </div>
+              <div v-if="currentAssignment.assignment.isStatic === false" class="mt-8 flex w-full items-center justify-end gap-6 px-10">
+                <button class="group flex items-center justify-center gap-2 rounded-xl bg-neutral-100 px-16 py-2 text-xl hover:bg-neutral-200" type="button" @click="submitQuestion()">Submit</button>
```
whats the () for

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -29,6 +30,15 @@
+              <div v-if="currentAssignment.assignment.isStatic === false" class="mt-8 flex w-full items-center justify-end gap-6 px-10">
+                <button class="group flex items-center justify-center gap-2 rounded-xl bg-neutral-100 px-16 py-2 text-xl hover:bg-neutral-200" type="button" @click="submitQuestion()">Submit</button>
+              </div>
+              <div v-if="feedbackMessage">
```
do these need to be in divs

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -29,6 +30,15 @@
+                <button class="group flex items-center justify-center gap-2 rounded-xl bg-neutral-100 px-16 py-2 text-xl hover:bg-neutral-200" type="button" @click="submitQuestion()">Submit</button>
+              </div>
+              <div v-if="feedbackMessage">
+                <p class="group flex items-center justify-center gap-2 rounded-xl px-16 py-2 text-xl text-neutral-400">{{ feedbackMessage }}</p>
```
whats group for

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -48,6 +58,19 @@ const route = useRoute();
 const router = useRouter();
 const userStore = useUserStore();
 const { studentCurrentCourse, currentQuestion } = storeToRefs(userStore);
+const feedbackMessage = ref<string>("");
```
explicit types
expect a lawsuit soon

rylan-geykhman replied:
sorry sir i plead guilty to not remembering your blog

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -48,6 +58,19 @@ const route = useRoute();
+const errorMessage = ref<string>("");
+const isAnswerCorrect = ref<boolean>(false);
+const remainingAttempts = ref<number>(0);
+const selectedChoice = ref<Question["answers"][0]>();
```
make a new type for this in types.ts called Answer or smth and export it
and change Questions to use the new type

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -48,6 +58,19 @@ const route = useRoute();
+const selectedChoice = ref<Question["answers"][0]>();
+
+function selectChoice(choice: Question["answers"][0]) {
+  currentQuestion.value?.question.answers.forEach((answer) => {
```
1 liner it

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -69,17 +92,24 @@ watch(
     if (!currentAssignment.value) return;
 
-    const alreadyFetchedQuestion = currentAssignment.value.assignment.questionInterfaces[currentQuestionIndex.value] as QuestionInterface | undefined;
+    const alreadyFetchedQuestion = currentAssignment.value.assignment.questionInterfaces[currentQuestionIndex.value] as QuestionInterface;
```
whats the reason for removing undefined

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
-  if (direction === "next" && currentQuestionIndex.value < currentAssignment.value.assignment.numQuestions - 1) await changeRouteQuery({ q: currentQuestionIndex.value + 1 });
+  const newIndex = direction === "previous" ? currentQuestionIndex.value - 1 : currentQuestionIndex.value + 1;
+
+  if (newIndex >= 0 && newIndex < currentAssignment.value.assignment.numQuestions) {
```
1 liner it

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
+
+async function submitQuestion() {
+  try {
+    if (!selectedChoice.value || !currentQuestion.value) {
```
remove console warn and 1 liner it

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -69,17 +92,24 @@ watch(
-      currentQuestion.value = question;
+      if (currentAssignment.value.assignment.isStatic) {
+        const question = await getNextStaticQuestion(currentAssignment.value.id, currentQuestionIndex.value + 1);
+        currentAssignment.value.assignment.questionInterfaces[currentQuestionIndex.value] = question;
```
this line and the line under it are repeated in both if else blocks, is there a way to not do that

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
+    } else if (response.remainingAttempts === 0) {
+      feedbackMessage.value = "You've exceeded the maximum amount of attempts on the previous question. It has been marked incorrect.";
+    } else {
+      feedbackMessage.value = `Incorrect. You have ${response.remainingAttempts ?? "infinite"} attempts left.`;
```
if they have infinite attempts just dont say how many attempts they have left

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
+    isAnswerCorrect.value = response.isCorrect;
+    remainingAttempts.value = response.remainingAttempts;
+
+    if (response.isCorrect) {
```
1 liner these

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
+    }
+
+    if (response.isCorrect || response.remainingAttempts === 0) {
+      await switchQuestion("next");
```
does this need to be awaited

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
+      feedbackMessage.value = `Incorrect. You have ${response.remainingAttempts ?? "infinite"} attempts left.`;
+    }
+
+    if (response.isCorrect || response.remainingAttempts === 0) {
```
1 liner it

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
+      }
+    }
+  },
+  { deep: true }
```
what is this deep for
currentQuestionIndex is an index

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
+watch(
+  currentQuestionIndex,
+  async () => {
+    if (currentAssignment.value?.assignment.isStatic) {
```
do you need 2 separate if statements

rylan-geykhman replied:
doesn't seem to work without 🤷

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -88,10 +118,56 @@ watch(
+  async () => {
+    if (currentAssignment.value?.assignment.isStatic) {
+      try {
+        if (selectedChoice.value && currentQuestion.value) {
```
1 liner

kennething commented on utils/fetch.ts:
```diff
@@ -42,14 +42,19 @@ export async function getCourseStudents(courseId: number) {
-export async function getNextQuestion(assignmentId: number) {
-  const data = await requestEndpoint<QuestionInterface>("courses/student/get-next-question/", "POST", { id: assignmentId });
+export async function getNextDynamicQuestion(assignmentId: number) {
+  /*   const data = await requestEndpoint<QuestionInterface>("courses/student/get-next-dynamic-question/", "POST", { id: assignmentId });
```
remove comments

kennething commented on utils/fetch.ts:
```diff
@@ -42,14 +42,19 @@ export async function getCourseStudents(courseId: number) {
 
-export async function getNextQuestion(assignmentId: number) {
-  const data = await requestEndpoint<QuestionInterface>("courses/student/get-next-question/", "POST", { id: assignmentId });
+export async function getNextDynamicQuestion(assignmentId: number) {
```
add jsdoc

kennething commented on utils/fetch.ts:
```diff
@@ -42,14 +42,19 @@ export async function getCourseStudents(courseId: number) {
+  return requestEndpoint<QuestionInterface>("courses/student/get-next-dynamic-question/", "POST", { id: assignmentId });
+}
+
+export async function getNextStaticQuestion(assignmentId: number, questionIndex: number) {
```
add jsdoc

kennething commented on utils/types.ts:
```diff
@@ -53,6 +55,7 @@ interface Assignment {
 }
 
 export interface StudentAssignment extends Assignment {
+  isStatic: boolean;
```
jsdoc

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -1,25 +1,26 @@
+                <button
+                  type="button"
+                  class="w-full rounded-lg bg-neutral-300 px-6 py-3 shadow-sm"
+                  :class="{
```
why is this 3 lines

kennething commented on pages/student/course/[courseCode]/[assignmentId]/index.vue:
```diff
@@ -29,6 +30,15 @@
                   <img class="size-5 group-hover:translate-x-1" src="/ui/arrowRight.svg" aria-hidden="true" />
                 </button>
               </div>
+              <div v-if="currentAssignment.assignment.isStatic === false" class="mt-8 flex w-full items-center justify-end gap-6 px-10">
```
yourHosIsGood === false

kennething committed:
Merge branch 'dev' into 300-feature-finish-question-page-requests

rylan-geykhman committed:
some fixes, not all

rylan-geykhman committed:
fixing typecheck error

rylan-geykhman committed:
kenneth fixes done, highlighting broken

- something something highlighting not particularly working
- should be a fix just something to do with bad logic

rylan-geykhman committed:
highlighting fixed

rylan-geykhman committed:
fixing typecheck

rylan-geykhman committed:
ternary update

kennething reviewed (approved):
LGTM

kennething committed:
Merge branch 'dev' into 300-feature-finish-question-page-requests
